core/src/mast/decorator_storage.rs

```diff
@@ -0,0 +1,1135 @@
+    /// * `operations_capacity` - Expected total number of operations across all nodes
+    /// * `decorator_ids_capacity` - Expected total number of decorator IDs across all operations
+    pub fn with_capacity(
+        nodes_capacity: usize,
+        operations_capacity: usize,
+        decorator_ids_capacity: usize,
+    ) -> Self {
+        Self {
+            decorator_ids: Vec::with_capacity(decorator_ids_capacity),
+            op_indptr_for_dec_ids: Vec::with_capacity(operations_capacity + 1),
+            node_indptr_for_op_idx: IndexVec::with_capacity(nodes_capacity + 1),
+        }
+    }
+
+    /// Create a new empty DecoratorIndexMapping.
+    pub fn new() -> Self {
+        Self::with_capacity(0, 0, 0)
+    }
+
+    /// Create a DecoratorIndexMapping from raw CSR components.
+    ///
+    /// This is useful for deserialization or testing purposes.
+    ///
+    /// # Arguments
+    /// * `decorator_ids` - Flat storage of all decorator IDs
+    /// * `op_indptr_for_dec_ids` - Pointer indices for operations within decorator_ids
+    /// * `node_indptr_for_op_idx` - Pointer indices for nodes within op_indptr_for_dec_ids
+    ///
+    /// # Returns
+    /// An error if the internal structure is inconsistent.
+    pub fn from_components(
+        decorator_ids: Vec<DecoratorId>,
+        op_indptr_for_dec_ids: Vec<usize>,
+        node_indptr_for_op_idx: IndexVec<MastNodeId, usize>,
+    ) -> Result<Self, DecoratorIndexError> {
+        // Validate the structure
+        if op_indptr_for_dec_ids.is_empty() {
+            return Err(DecoratorIndexError::InternalStructure);
+        }
+
+        // Check that the last operation pointer doesn't exceed decorator IDs length
+        let Some(&last_op_ptr) = op_indptr_for_dec_ids.last() else {
+            return Err(DecoratorIndexError::InternalStructure);
+        };
+        if last_op_ptr > decorator_ids.len() {
+            return Err(DecoratorIndexError::InternalStructure);
+        }
+
+        // Check that node pointers are within bounds of operation pointers
+        if node_indptr_for_op_idx.is_empty() {
+            return Err(DecoratorIndexError::InternalStructure);
+        }
+
+        let node_slice = node_indptr_for_op_idx.as_slice();
+        let Some(&last_node_ptr) = node_slice.last() else {
+            return Err(DecoratorIndexError::InternalStructure);
+        };
+        if last_node_ptr > op_indptr_for_dec_ids.len() - 1 {
+            return Err(DecoratorIndexError::InternalStructure);
+        }
+
+        // Ensure monotonicity of pointers
+        for window in op_indptr_for_dec_ids.windows(2) {
+            if window[0] > window[1] {
+                return Err(DecoratorIndexError::InternalStructure);
+            }
+        }
+
+        for window in node_slice.windows(2) {
+            if window[0] > window[1] {
+                return Err(DecoratorIndexError::InternalStructure);
+            }
+        }
+
+        Ok(Self {
```

Comment:
Q: aren't there some restrictions on admissible inputs to this function? I am thinking not per component but between components

Reply:
Yes, good remark, I've made its purpose implicit by making it test-only.